navigator/runtime-experimental/src/main/java/com/freeletics/mad/navigator/compose/internal/MultiStack.kt

```diff
@@ -0,0 +1,181 @@
+package com.freeletics.mad.navigator.compose.internal
+
+import android.os.Bundle
+import androidx.compose.runtime.MutableState
+import androidx.compose.runtime.State
+import androidx.compose.runtime.mutableStateOf
+import androidx.core.os.bundleOf
+import com.freeletics.mad.navigator.BaseRoute
+import com.freeletics.mad.navigator.NavRoot
+import com.freeletics.mad.navigator.NavRoute
+import com.freeletics.mad.navigator.compose.ContentDestination
+import com.freeletics.mad.navigator.internal.DestinationId
+import com.freeletics.mad.navigator.internal.destinationId
+import java.util.UUID
+
+internal class MultiStack(
+    private val allStacks: MutableList<Stack>,
+    private val startStack: Stack,
+    private var currentStack: Stack,
+    private val destinations: List<ContentDestination<*>>,
+    private val onStackEntryRemoved: (StackEntry.Id) -> Unit,
+    private val idGenerator: () -> String,
+) {
+
+    private val visibleEntryState: MutableState<List<StackEntry<*>>> =
+        mutableStateOf(currentStack.computeVisibleEntries())
+    val visibleEntries: State<List<StackEntry<*>>>
+        get() = visibleEntryState
+
+    private val canNavigateBackState: MutableState<Boolean> =
+        mutableStateOf(canNavigateBack())
+    val canNavigateBack: State<Boolean>
+        get() = canNavigateBackState
+
+    val startRoot = startStack.rootEntry.route as NavRoot
+
+    fun <T : BaseRoute> entryFor(destinationId: DestinationId<T>): StackEntry<T>? {
+        val entry = currentStack.entryFor(destinationId)
+        if (entry != null) {
+            return entry
+        }
+
+        // the root of the default back stack is always on the back stack
+        if (startStack.rootEntry.destinationId == destinationId) {
+            @Suppress("UNCHECKED_CAST")
+            return startStack.rootEntry as StackEntry<T>
+        }
+
+        return null
+    }
+
+    private fun getOrCreateBackStack(root: NavRoot): Stack {
+        val existingStack = allStacks.find { it.id == root.destinationId }
+        if (existingStack != null) {
+            return existingStack
+        }
+
+        val newStack = Stack.createWith(root, destinations, onStackEntryRemoved, idGenerator)
+        allStacks.add(newStack)
+        return newStack
+    }
+
+    private fun switchToBackStack(
+        stack: Stack,
+        restoreRootState: Boolean,
+        saveCurrentRootState: Boolean
```

Comment:
I think `removeCurrentRootState` would be clearer here since it emphasizes positive action instead of `saveCurrentRootState` which is only does something when it's `false`. 

Similar thing about `restoreRootState` since it's not about restoring but rather not clearing. Perhaps `clearNewRootState` would work better.